src/wazuh_modules/syscollector/sysOsParsers.cpp

```diff
@@ -0,0 +1,402 @@
+{
+    bool ret{false};
+    if (start.empty() || Utils::startsWith(in, start))
+    {
+        std::smatch sm;
+        ret = std::regex_search(in, sm, pattern);
+        if (ret && sm.size() >= matchIndex)
+        {
+            match = sm[matchIndex];
+        }
+    }
+    return ret;
+}
+
+static bool findCodeNameInString(const std::string& in,
+                                 std::string& output)
+{
+    const auto end{in.rfind(")")};
+    const auto start{in.rfind("(")};
+    const bool ret
+    {
+        start != std::string::npos && end != std::string::npos
+    };
+    if (ret)
+    {
+        output = Utils::trim(in.substr(start + 1, end - (start + 1)));
+    }
+    return ret;
+}
+
+static bool findMajorMinorVersionInString(const std::string& in,
+                                         nlohmann::json& output)
+{
+    constexpr auto MINOR_VERSION_PATTERN{"^[0-9]+\\.([0-9]+)\\.*"};
+    constexpr auto MAJOR_VERSION_PATTERN{"^([0-9]+)\\.*"};
+    std::string version;
+    std::regex pattern{MAJOR_VERSION_PATTERN};
+    const auto ret{findRegexInString(in, version, pattern, 1)};
+    if (ret)
+    {
+        output["os_major"] = version;
+    }
+    pattern = MINOR_VERSION_PATTERN;
+    if (findRegexInString(in, version, pattern, 1))
+    {
+        output["os_minor"] = version;
+    }
+    return ret;
+}
+
+
+static bool findVersionInStream(std::istream& in,
+                                nlohmann::json& output,
+                                const std::string& regex,
+                                const size_t matchIndex = 0,
+                                const std::string& start = "")
+{
+    bool ret{false};
+    std::string line;
+    std::string data;
+    std::regex pattern{regex};
+    while(std::getline(in, line))
+    {
+        line = Utils::trim(line);
+        ret |= findRegexInString(line, data, pattern, matchIndex, start);
+        if (ret)
+        {
+            output["os_version"] = data;
+            findMajorMinorVersionInString(data, output);
+        }
+        if (findCodeNameInString(line, data))
+        {
+            output["os_codename"] = data;
+        }
+    }
+    return ret;
+}
+
+bool UnixOsParser::parseFile(std::istream& in, nlohmann::json& info)
+{
+    constexpr auto SEPARATOR{'='};
+    static const std::map<std::string, std::string> KEY_MAP
+    {
+        {"NAME",             "os_name"},
+        {"VERSION",          "os_version"},
+        {"ID",               "os_platform"},
+        {"VERSION_CODENAME", "os_codename"}
+    };
+    const auto ret {parseUnixFile(KEY_MAP, SEPARATOR, in, info)};
+    if (ret)
+    {
+        findMajorMinorVersionInString(info["os_version"], info);
+    }
+    return ret;
+}
+
+bool UbuntuOsParser::parseFile(std::istream& in, nlohmann::json& output)
+{
+    constexpr auto PATTERN_MATCH{R"([0-9].*\.[0-9]*)"};
+    constexpr auto DISTRIB_FIELD{"DISTRIB_DESCRIPTION"};
+    static const std::string CODENAME_FIELD{"DISTRIB_CODENAME"};
```

Comment:
constexpr auto ?

Reply:
if I use a constexpr I'll need to call strlen on the string literal on every call to parseFile or create a static size_t with the value having then 2 const variables instead of just 1. That's why I preferred the static const std::string.